src/Features/Core/Portable/CodeFixesAndRefactorings/CodeActionRequestPriorityProvider.cs

```diff
@@ -0,0 +1,107 @@
+        {
+            if (priority == CodeActionRequestPriority.None)
+                return Default;
+
+            return MutableProvider.Create(priority);
+        }
+
+        /// <summary>
+        /// Returns true if the given <paramref name="analyzer"/> can report diagnostics that can have
+        /// fixes from a code fix provider with <see cref="CodeFixProvider.RequestPriority"/>
+        /// matching <see cref="Priority"/>.
+        /// This method is useful for performing a performance optimization for lightbulb diagnostic computation,
+        /// wherein we can reduce the set of analyzers to be executed when computing fixes for a specific
+        /// <see cref="Priority"/>.
+        /// </summary>
+        public bool MatchesPriority(DiagnosticAnalyzer analyzer)
+        {
+            // If caller isn't asking for prioritized result, then run all analyzers.
+            if (Priority == CodeActionRequestPriority.None)
+                return true;
+
+            // 'CodeActionRequestPriority.Lowest' is used for suppression/configuration fixes,
+            // which requires all analyzer diagnostics.
+            if (Priority == CodeActionRequestPriority.Lowest)
+                return true;
+
+            // The compiler analyzer always counts for any priority.  It's diagnostics may be fixed
+            // by high pri or normal pri fixers.
+            if (analyzer.IsCompilerAnalyzer())
+                return true;
+
+            // Check if we are computing diagnostics for 'CodeActionRequestPriority.Low' and
+            // this analyzer was de-prioritized to low priority bucket.
+            if (Priority == CodeActionRequestPriority.Low &&
+                IsDeprioritizedAnalyzerWithLowPriority(analyzer))
+            {
+                return true;
+            }
+
+            var analyzerPriority = analyzer is IBuiltInAnalyzer { RequestPriority: var requestPriority }
+                ? requestPriority
+                : CodeActionRequestPriority.Normal;
+
+            return Priority == analyzerPriority;
+        }
+
+        /// <summary>
+        /// Returns true if the given <paramref name="codeFixProvider"/> should be considered
+        /// a candidate when computing fixes for the given <see cref="Priority"/>.
+        /// </summary>
+        public bool MatchesPriority(
+            CodeFixProvider codeFixProvider)
+            => Priority switch
+            {
+                // We are computing fixes for all priorities
+                CodeActionRequestPriority.None => true,
+
+                // 'Low' priority is used for fixes for expensive analyzers which were de-prioritized
+                // to low priority bucket.
+                // We accept fixers with any RequestPriority that can fix diagnostics
+                // from low priority analyzers.
+                CodeActionRequestPriority.Low => true,
+
+                _ => Priority == codeFixProvider.RequestPriority,
+            };
```

Comment:
something seems super off about this method.  specifically, i can't wrap my head around the .Low case yet.

Reply:
So the reason why we need this handling for `Low` bucket is that for a de-prioritized analyzer, the code fixer itself has `Normal` priority, and even the analyzer declares it's request priority as `Normal` by default. Because the engine pushed the analyzer to `Low` bucket, we need to allow code fixes that claim to be needed only for `Normal` pri bucket to be allowed in `Low` pri bucket.

Reply:
I have added more details in the comment.